nupic/research/frameworks/dendrites/modules/dendritic_mlp.py

```diff
@@ -67,83 +72,101 @@ class DendriticMLP(nn.Module):
     """
 
-    def __init__(self, input_size, output_size, hidden_size, num_segments, dim_context,
-                 kw=True, dendrite_sparsity=0.95, weight_init="modified",
-                 dendrite_init="modified", freeze_dendrites=False,
-                 dendritic_layer_class=AbsoluteMaxGatingDendriticLayer):
+    def __init__(
+        self, input_size, output_size, hidden_sizes, num_segments, dim_context,
+        kw, kw_percent_on=0.05, context_percent_on=1.0, weight_sparsity=0.95,
+        weight_init="modified", dendrite_init="modified", freeze_dendrites=False,
+        dendritic_layer_class=AbsoluteMaxGatingDendriticLayer, output_nonlinearity=None
+    ):
 
         # Forward & dendritic weight initialization must be either "kaiming" or
         # "modified"
         assert weight_init in ("kaiming", "modified")
         assert dendrite_init in ("kaiming", "modified")
+        assert kw_percent_on >= 0.0
+        assert context_percent_on >= 0.0
 
         super().__init__()
 
+        if num_segments == 1:
+            # use optimized 1 segment class
+            dendritic_layer_class = OneSegmentDendriticLayer
+
         self.num_segments = num_segments
         self.input_size = input_size
-        self.hidden_size = hidden_size
+        self.hidden_sizes = hidden_sizes
         self.output_size = output_size
         self.dim_context = dim_context
         self.kw = kw
-        self.dendrite_sparsity = dendrite_sparsity
-        self.freeze_dendrites = freeze_dendrites
-
-        # Forward layers & k-winners
-        self.dend1 = dendritic_layer_class(
-            module=nn.Linear(input_size, hidden_size, bias=True),
-            num_segments=num_segments,
-            dim_context=dim_context,
-            module_sparsity=0.95,
-            dendrite_sparsity=dendrite_sparsity,
-        )
-        self.dend2 = dendritic_layer_class(
-            module=nn.Linear(hidden_size, hidden_size, bias=True),
-            num_segments=num_segments,
-            dim_context=dim_context,
-            module_sparsity=0.95,
-            dendrite_sparsity=dendrite_sparsity,
-        )
-        self.classifier = SparseWeights(module=nn.Linear(hidden_size, output_size),
-                                        sparsity=0.95)
-
-        if kw:
-
-            print(f"Using k-Winners: 0.05 'on'")
-            self.kw1 = KWinners(n=hidden_size, percent_on=0.05, k_inference_factor=1.0,
-                                boost_strength=0.0, boost_strength_factor=0.0)
-            self.kw2 = KWinners(n=hidden_size, percent_on=0.05, k_inference_factor=1.0,
-                                boost_strength=0.0, boost_strength_factor=0.0)
-
+        self.kw_percent_on = kw_percent_on
+        self.weight_sparsity = weight_sparsity
+        self.output_nonlinearity = output_nonlinearity
+        self.hardcode_dendrites = (dendrite_init == "hardcoded")
+
+        self._layers = nn.ModuleList()
+        self._activations = nn.ModuleList()
```

Comment:
Have you considered just wrapping the whole dendritic MLP in a `torch.nn.Sequential`? That will simplify the `forward` call. I just want to point this out, but it's up to you.